pkg/services/authz/rbac/store/folder_store.go

```diff
@@ -0,0 +1,181 @@
+	"context"
+	"fmt"
+
+	"github.com/grafana/authlib/types"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+
+	"github.com/grafana/grafana/pkg/apimachinery/utils"
+	folderv0 "github.com/grafana/grafana/pkg/apis/folder/v0alpha1"
+	"github.com/grafana/grafana/pkg/infra/tracing"
+	"github.com/grafana/grafana/pkg/storage/legacysql"
+	"github.com/grafana/grafana/pkg/storage/unified/sql/sqltemplate"
+)
+
+type FolderStore interface {
+	ListFolders(ctx context.Context, ns types.NamespaceInfo) ([]Folder, error)
+}
+
+type Folder struct {
+	UID       string
+	ParentUID *string
+}
+
+var _ FolderStore = (*SQLFolderStore)(nil)
+
+func NewSQLFolderStore(sql legacysql.LegacyDatabaseProvider, tracer tracing.Tracer) *SQLFolderStore {
+	return &SQLFolderStore{sql, tracer}
+}
+
+type SQLFolderStore struct {
+	sql    legacysql.LegacyDatabaseProvider
+	tracer tracing.Tracer
+}
+
+var sqlFolders = mustTemplate("folder_query.sql")
+
+type listFoldersQuery struct {
+	sqltemplate.SQLTemplate
+
+	Query       *FolderQuery
+	FolderTable string
+}
+
+type FolderQuery struct {
+	OrgID int64
+}
+
+func (r listFoldersQuery) Validate() error {
+	return nil
+}
+
+func newListFolders(sql *legacysql.LegacyDatabaseHelper, query *FolderQuery) listFoldersQuery {
+	return listFoldersQuery{
+		SQLTemplate: sqltemplate.New(sql.DialectForDriver()),
+		Query:       query,
+		FolderTable: sql.Table("folder"),
+	}
+}
+
+func (s *SQLFolderStore) ListFolders(ctx context.Context, ns types.NamespaceInfo) ([]Folder, error) {
+	ctx, span := s.tracer.Start(ctx, "authz_direct_db.database.ListFolders")
+	defer span.End()
+
+	sql, err := s.sql(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	query := newListFolders(sql, &FolderQuery{OrgID: ns.OrgID})
+	q, err := sqltemplate.Execute(sqlFolders, query)
+	if err != nil {
+		return nil, err
+	}
+
+	rows, err := sql.DB.GetSqlxSession().Query(ctx, q, query.GetArgs()...)
+	defer func() {
+		if rows != nil {
+			_ = rows.Close()
+		}
+	}()
+	if err != nil {
+		return nil, err
+	}
+
+	var folders []Folder
+	for rows.Next() {
+		var folder Folder
+		if err := rows.Scan(&folder.UID, &folder.ParentUID); err != nil {
+			return nil, err
+		}
+		folders = append(folders, folder)
+	}
+
+	return folders, nil
+}
+
+var _ FolderStore = (*APIFolderStore)(nil)
+
+func NewAPIFolderStore(tracer tracing.Tracer, configProvider func(ctx context.Context) *rest.Config) *APIFolderStore {
+	return &APIFolderStore{tracer, configProvider}
+}
+
+type APIFolderStore struct {
+	tracer         tracing.Tracer
+	configProvider func(ctx context.Context) *rest.Config
+}
+
+func (s *APIFolderStore) ListFolders(ctx context.Context, ns types.NamespaceInfo) ([]Folder, error) {
+	ctx, span := s.tracer.Start(ctx, "authz.apistore.ListFolders")
+	defer span.End()
+
+	client, err := s.client(ctx, ns.Value)
+	if err != nil {
+		return nil, fmt.Errorf("create resource client: %w", err)
+	}
+	list := func(c string) ([]Folder, string, error) {
```

Comment:
I think client-go has a [Pager](https://github.com/kubernetes/client-go/blob/master/tools/pager/pager.go#L48) able to buffer pages

Reply:
Cool, will take a look!

Reply:
Fixed in [c16a643](https://github.com/grafana/grafana/pull/100038/commits/c16a643efd5964abfceab73ea1b069ad13a0160f)

Reply:
There was also a import cycle in tests

apiserver -> authz -> sql -> apistore -> apiserver. So I had to fix tests in apistore [98199c6](https://github.com/grafana/grafana/pull/100038/commits/98199c6f64ed386c3666436b0fcd2584e1ae61ec)